monkey/common/di_container.py

```diff
@@ -0,0 +1,122 @@
+        """
+        Register a concrete instance that satisfies a given interface.
+
+        :param interface: An interface or abstract base class that other classes depend upon
+        :param instance: An instance (object) of a type that implements `interface`
+        """
+        self._instance_registry[interface] = instance
+        DIContainer._del_key(self._type_registry, interface)
+
+    def resolve(self, type_: Type[T]) -> T:
+        """
+        Resolves all dependencies and returns a new instance of `type_` using constructor dependency
+        injection. Note that only positional arguments are resolved. Varargs, keyword-only args, and
+        default values are ignored.
+
+        :param type_: A type (class) to construct
+        :return: An instance of `type_`
+        """
+        try:
+            return self._resolve_type(type_)
+        except UnregisteredTypeError:
+            pass
+
+        args = self.resolve_dependencies(type_)
+        return type_(*args)
+
+    def resolve_dependencies(self, type_: Type[T]) -> Sequence[Any]:
+        """
+        Resolves all dependencies of type_ and returns a Sequence of objects that correspond type_'s
+        dependencies. Note that only positional arguments are resolved. Varargs, keyword-only args,
+        and default values are ignored.
+
+        :param type_: A type (class) to resolve dependencies for
+        :return: An Sequence of dependencies to be injected into type_'s constructor
+        """
+        args = []
+
+        for arg_type in inspect.getfullargspec(type_).annotations.values():
+            instance = self._resolve_type(arg_type)
+            args.append(instance)
+
+        return tuple(args)
+
+    def _resolve_type(self, type_: Type[T]) -> T:
+        if type_ in self._type_registry:
+            return self._construct_new_instance(type_)
+        elif type_ in self._instance_registry:
+            return self._retrieve_registered_instance(type_)
+
+        raise UnregisteredTypeError(f'Failed to resolve unregistered type "{type_.__name__}"')
+
+    def _construct_new_instance(self, arg_type: Type[T]) -> T:
+        try:
+            return self._type_registry[arg_type]()
+        except TypeError:
+            # arg_type has dependencies that must be resolved. Recursively call resolve() to
+            # construct an instance of arg_type with all of the requesite dependencies injected.
+            return self.resolve(self._type_registry[arg_type])
+
+    def _retrieve_registered_instance(self, arg_type: Type[T]) -> T:
+        return self._instance_registry[arg_type]
+
+    def release(self, interface: Type[T]):
+        """
+        Deregister's an interface
+
+        :param interface: The interface to release
+        """
+        DIContainer._del_key(self._type_registry, interface)
+        DIContainer._del_key(self._instance_registry, interface)
+
+    @staticmethod
+    def _del_key(mapping: MutableMapping[T, Any], key: T):
```

Comment:
Perhaps we should establish a `try_do_something` naming convention for all error handling thin wrappers.

Reply:
`try` implies that it will raise an exception, which is exactly the opposite of what this does.